[gen] normalized fields with annotations holding original field name

hochgi opened the pull request:
PR ready for review

fixes #3031 
/claim #3031

hochgi committed:
[gen] clean some generated code annotations boilerplate in preparation for #3031 (which will add more annotations). Specifically: use imports insteads of inlined FQCN when adding annotations

hochgi commented on zio-http-gen/src/test/scala/zio/http/gen/scala/CodeGenSpec.scala:
```diff
@@ -771,5 +771,5 @@ object CodeGenSpec extends ZIOSpecDefault {
         }
       },
-    ) @@ java11OrNewer @@ flaky @@ blocking // Downloading scalafmt on CI is flaky
+    ) @@ java11OrNewer /*@@ flaky*/ @@ blocking // Downloading scalafmt on CI is flaky
```
TODO: don't forget to re-enable

hochgi committed:
* add zio-parser & zio-config dependencies to zio-http-gen module
* add new configurations needed for normalization of fields

hochgi committed:
* ditch zio-parser and use a complex regex that does the same (more or less), since previous implementation was slooooooooow.
* propagate configuration to use fields normalization
* implement & add relevant annotation when field name been normalized
* test suite for normalization logic
* add test to verify annotations are add correctly

hochgi commented on zio-http-gen/src/test/resources/inline_schema_weird_field_names.yaml:
this file is C&Ped from a different test, and I only test the order class - I should ditch large parts of its less relevant content.

hochgi commented on zio-http-gen/src/test/resources/ComponentOrderWithAliases.scala:
```diff
@@ -1,12 +1,14 @@
-  @zio.schema.annotation.validate[Int](zio.schema.validation.Validation.greaterThan(0)) quantity: Int,
-  @zio.schema.annotation.validate[Double](zio.schema.validation.Validation.greaterThan(-1.0)) price: Double,
+  @validate[Int](Validation.greaterThan(0)) quantity: Int,
+  @validate[Double](Validation.greaterThan(-1.0)) price: Double,
```
Not related to current PR, but the float/double validation logic is faulty.
`minimum 0` does not mean `> -1.0`
Since e.g: `-0.5 > -1.0`, but does not adhere to original minimum requirement.

CC @987Nabil

987Nabil replied:
True. But I think this is unsolvable then. There is no way to translate it to zio-schema correctly right now. We probably need to add new validations to zio schema

hochgi replied:
@987Nabil in second thought, perhaps you need a change for greaterOrEqualTo, but we can still use `java.lang.Math.nextDown` (or `nextAfter`) to get the closest representable floating point number that is smaller (or larger) than the input.

It might be ugly, e.g:
```sacla
@ java.lang.Math.nextDown(0f)
res0: Float = -1.4E-45F
```
but at least it'll be (effectively) correct.

WDYT?

987Nabil replied:
I think that works work. But I would not write out the number, but generate the expression

hochgi replied:
Yup...
```scala
@validate(Validation.greaterThan(Math.nextDown(0f)))
```
Would be much better (and understandable) than generating weird epsilons

hochgi committed:
inconsistent order of fields in different scala versions, so sorting fields (and imports) when rendering

hochgi committed:
adjusted test expectations to match sorted fields/imports change

hochgi committed:
more fields ordering fixes for tests

hochgi committed:
re-enable flaky

hochgi committed:
remove redundant stuff from test file

hochgi commented on project/Dependencies.scala:
```diff
@@ -36,7 +37,11 @@ object Dependencies {
+  val `zio-config-magnolia` = "dev.zio" %% "zio-config-magnolia" % ZioConfigVersion
+  val `zio-config-typesafe` = "dev.zio" %% "zio-config-typesafe" % ZioConfigVersion
   val `zio-json-yaml`       = "dev.zio" %% "zio-json-yaml"       % ZioJsonVersion
+  val `zio-parser`          = "dev.zio" %% "zio-parser"          % ZioParserVersion
```
zio-parser implementation wasn't good enough (parser was very slow - about a minute runtime per call. equivalent scala-parser-combinators implementation was fast).
I ended up replacing with a good enough regex (with the help of co-pilot).
This dependency is currently not used, and can be removed.

987Nabil replied:
Please remove it then. Btw, we have RichTextCodec, which might help.

hochgi commented on zio-http-gen/src/main/scala/zio/http/gen/openapi/Config.scala:
```diff
@@ -19,20 +23,55 @@ final case class Config(
+     * unless original field is defined in the specialReplacements map.
+     */,
+    specialReplacements: Map[String, String], /*
+     * When normalization is enabled, a heuristic parser will attempt to normalize field names.
```
Map can be used also when auto normalization is not enabled. I should amend the documentation to be clearer

987Nabil replied:
If `specialReplacements` can be used out side of normalization, does it belong in the normalization config?

hochgi replied:
I meant of course it's part of normalization. but it overrides **auto** normalization (i.e: using the parser to construct a different term).
Normalization has 2 modes:
- manual replacements (user control exactly what to replace with what)
- automatic (a heuristic parser will try to do the best it can)

both fall under normalization.
Perhaps I should rename as: `automaticEnabled` & `manualReplacements`?

987Nabil replied:
Would turning the automated off lead to invalid code? Should it be always on? And the manual just be tried first, if the map is not empty?

hochgi replied:
no, not at all.
These are decoupled.

having e.g. auto off, and replacement on can yield from e.g:
```yaml
Thing:
  type: object
  properties:
    S/N:
      type: string
      format: uuid
    N_A_M_E:
      type: String
```
with only replacing `"N_A_M_E" -> "name"` something like:
```scala
case class Thing (
  `S/N`: UUID,
  name: String
)
```

which is perfectly fine.

hochgi commented on zio-http-gen/src/main/scala/zio/http/gen/openapi/EndpointGen.scala:
```diff
@@ -729,14 +730,14 @@ final case class EndpointGen(config: Config) {
         .asInstanceOf[Code.Field]
       if (obj.required.contains(name)) field else field.copy(fieldType = field.fieldType.opt)
-    }.toList
+    }
```
the `obj.properties: Map[String, JsonSchema]` when converted to a list has inconsistent ordering in different Scala versions. Ideally, we should preserve the order of fields as appears originally in OpenAPI .
Not sure how to pull that off, or if its even worth the hassle.
Any suggestions @987Nabil ?

hochgi replied:
I opted for sorting by (original) name, which is at least consistent.

hochgi replied:
As discussed over discord (and to persist in a visible place), fields inconsistent ordering stem from zio-schema internally using a `Map` to represent a Dictionary:
https://github.com/zio/zio-schema/blob/c741f6b3b771d7e08c6936f9f54539735a71aa78/zio-schema/shared/src/main/scala/zio/schema/DynamicValue.scala#L95-L104

We've managed to avoid it so far, since almost all test cases are small (4 fields or less), and Map1, …, Map4 implementations across different scala versions remain consistent in term of element ordering.

When I used a test with 7 fields, it exposed the fact that real maps (> 4 elements) are inconsistent in different scala versions.

I tried replacing Map with ListMap, but the derived deserializer is already inconsistent.

@987Nabil suggests it may even be an issue in zio-json (is internal Json object relies on Map or Chunk[(String, Json)]?)

I ended up reducing the test to only 4 fields, and now it's consistent.
But the issue remains.

What's the risk?
Someone relies on generated code, and have same type fields, e.g: `case class Thing(a: String, b: String, …, e: String)`
by simply upgrading scala version, fields order may change, but code will still compile.
So having code like `Thing("A", "B", …, "E")` is now wrong and hard to track.

hochgi replied:
issue: https://github.com/zio/zio-schema/issues/736

987Nabil commented on zio-http-gen/src/main/scala/zio/http/gen/openapi/Config.scala:
```diff
@@ -19,20 +23,55 @@ final case class Config(
 object Config {
 
+  final case class NormalizeFields(
+    enabled: Boolean /*
```
Block comments belong before not after what is commented. Anyhow, I think this should be comments for the config class
```
  /**
     * 
     * @param enabled
     */
```

987Nabil commented on zio-http-gen/src/main/scala/zio/http/gen/openapi/Config.scala:
```diff
@@ -19,20 +23,55 @@ final case class Config(
   val default: Config = Config(
     commonFieldsOnSuperType = false,
     generateSafeTypeAliases = false,
+    fieldsNormalizationConf = NormalizeFields(
```
if you keep this variable, remove the conf from the name. We are in conf context anyway

hochgi committed:
* Undo fields sorting
* Restore test cases fields original order
* Reduced normalized test to have only 4 fields (up to 4 fields, order is consistent, since Map1, …, Map4 implementation is consistent between scala versions. More fields aren't guaranteed to have consistent order)

hochgi committed:
fmt

hochgi committed:
* removed zio-parser dependency
* renamed config normalizations setting to be clearer
* moved comments to be scaladoc in configuration (scalafmt kep messing this up, so I've set '//format: off' - should we configure 'docstrings.style = keep' in scalafmt.conf to avoid it messing up nicely formatted scaladocs?)

hochgi committed:
CI failed for docusaurus step, perhaps I should put scalafmt escaping on different lines?

hochgi committed:
3rd field was dropped from content, was redundant in config.
also, re-run CI since seems to be affected by a flaky test:
Client / ClientSpec / request can be timed out manually while awaiting connection
